storage/src/adb/any/variable/sync.rs

```diff
@@ -213,16 +495,269 @@ async fn compute_offset<E: Storage + Metrics, V: Codec>(
     Ok((current_offset as u64) * ITEM_ALIGNMENT)
 }
 
+/// Count the actual number of items in a journal section.
+async fn count_items_in_section<E: Storage + Metrics, V: Codec>(
+    blob: &commonware_runtime::buffer::Append<E::Blob>,
+    codec_config: &V::Cfg,
+    compressed: bool,
+) -> Result<u32, crate::journal::Error> {
+    let mut current_offset = 0u32;
+    let mut item_count = 0u32;
+
+    loop {
+        match VJournal::<E, V>::read(compressed, codec_config, blob, current_offset).await {
+            Ok((next_slot, _item_len, _item)) => {
+                current_offset = next_slot;
+                item_count += 1;
+            }
+            Err(crate::journal::Error::Runtime(
+                commonware_runtime::Error::BlobInsufficientLength,
+            )) => {
+                // Reached the end of the section
+                break;
+            }
+            Err(e) => return Err(e),
+        }
+    }
+
+    Ok(item_count)
+}
+
+/// Remove items before the `lower_bound` location from the first section of the journal, if any.
+/// If the section containing `lower_bound` has elements before `lower_bound`, these are removed.
+/// The remaining elements are then replayed at the beginning of the section.
+async fn prune_lower<E, V>(
+    journal: &mut VJournal<E, V>,
+    lower_bound: u64,
+    items_per_section: NonZeroU64,
+    oldest_retained_loc: u64,
+) -> Result<(), crate::journal::Error>
+where
+    E: Storage + Metrics + Clock,
+    V: Codec,
+{
+    // Find which section contains the lower_bound item
+    let lower_section = lower_bound / items_per_section.get();
+
+    let Some(_blob) = journal.blobs.get(&lower_section) else {
+        return Ok(()); // Section doesn't exist, nothing to prune
+    };
+
+    debug!(
+        lower_section,
+        lower_bound, oldest_retained_loc, "rebuilding section to remove items before lower_bound"
+    );
+
+    // Read all operations from the current section
+    let mut operations_to_keep = Vec::new();
```

Comment:
Given we now have metadata, it seems like we can avoid this rewrite by just storing a "first valid" key (that we wait for on restart)?

Reply:
Rewrote to only physically rewrite the lower section in the event that its elements are non-contiguous with the new lower_bound.